# Import Libraries

In [1]:
import sqlite3
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from math import ceil

# for better resolution plots
%config InlineBackend.figure_format = 'retina'

#o svg consegue ampliar infinitamente os gráficos sem perder qualidade mas às vezes é mais lento 
#por isso agora usamos retina

sns.set()

# Import Dataset

In [2]:
flightDB = pd.read_csv('DM_AIAI_FlightsDB.csv', sep = ",")
customerDB = pd.read_csv('DM_AIAI_CustomerDB.csv', sep = ",")

# General Look at the Flight DataSet

In [ ]:
flightDB.shape

(608436, 10)
(16921, 21)


In [ ]:
flightDB.head(10)


   Loyalty#  Year  Month YearMonthDate  NumFlights  NumFlightsWithCompanions  \
0    413052  2021     12     12/1/2021         2.0                       2.0   
1    464105  2021     12     12/1/2021         0.0                       0.0   
2    681785  2021     12     12/1/2021        10.0                       3.0   
3    185013  2021     12     12/1/2021        16.0                       4.0   
4    216596  2021     12     12/1/2021         9.0                       0.0   
5    486956  2021     12     12/1/2021        12.0                       7.0   
6    247514  2021     12     12/1/2021        17.0                       7.0   
7    711864  2021     12     12/1/2021         6.0                       0.0   
8    721372  2021     12     12/1/2021        11.0                       3.0   
9    762715  2021     12     12/1/2021         0.0                       0.0   

   DistanceKM  PointsAccumulated  PointsRedeemed  DollarCostPointsRedeemed  
0      9384.0              938.0          

In [70]:
flightDB.tail(10)

Loyalty#  Year  Month YearMonthDate  NumFlights  \
608426    999631  2019     12     12/1/2019         3.6   
608427    999731  2019     12     12/1/2019         0.0   
608428    999758  2019     12     12/1/2019         0.0   
608429    999788  2019     12     12/1/2019         0.0   
608430    999891  2019     12     12/1/2019         0.0   
608431    999902  2019     12     12/1/2019         7.2   
608432    999911  2019     12     12/1/2019         0.0   
608433    999940  2019     12     12/1/2019        14.4   
608434    999982  2019     12     12/1/2019         0.0   
608435    999986  2019     12     12/1/2019         0.0   

        NumFlightsWithCompanions  DistanceKM  PointsAccumulated  \
608426                       1.8     12262.5            1226.25   
608427                       0.0         0.0               0.00   
608428                       0.0         0.0               0.00   
608429                       0.0         0.0               0.00   
608430                       0.0         0.0               0.00   
608431                       0.0     30766.5            3076.65   
608432                       0.0         0.0               0.00   
608433                       0.9     18261.0            1826.10   
608434                       0.0         0.0               0.00   
608435                       0.0         0.0               0.00   

        PointsRedeemed  DollarCostPointsRedeemed  
608426             0.0                       0.0  
608427             0.0                       0.0  
608428             0.0                       0.0  
608429             0.0                       0.0  
608430             0.0                       0.0  
608431             0.0                       0.0  
608432             0.0                       0.0  
608433             0.0                       0.0  
608434             0.0                       0.0  
608435             0.0                       0.0

O NumFlights e NumFlightsWithCompanions têm decimais enquanto deviam ser integers
Os PointsAccumulated e PointsRedeemed devem ter decimais??

In [71]:
flightDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 608436 entries, 0 to 608435
Data columns (total 10 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Loyalty#                  608436 non-null  int64  
 1   Year                      608436 non-null  int64  
 2   Month                     608436 non-null  int64  
 3   YearMonthDate             608436 non-null  object 
 4   NumFlights                608436 non-null  float64
 5   NumFlightsWithCompanions  608436 non-null  float64
 6   DistanceKM                608436 non-null  float64
 7   PointsAccumulated         608436 non-null  float64
 8   PointsRedeemed            608436 non-null  float64
 9   DollarCostPointsRedeemed  608436 non-null  float64
dtypes: float64(6), int64(3), object(1)
memory usage: 46.4+ MB


Como todas as variáveis têm 608436 rows não nulas, está ok

In [72]:
flightDB.dtypes

Loyalty#                      int64
Year                          int64
Month                         int64
YearMonthDate                object
NumFlights                  float64
NumFlightsWithCompanions    float64
DistanceKM                  float64
PointsAccumulated           float64
PointsRedeemed              float64
DollarCostPointsRedeemed    float64
dtype: object

Esta info pode estar errada consoante o tipo de variáveis

In [73]:
flightDB.describe()

Loyalty#           Year          Month     NumFlights  \
count  608436.000000  608436.000000  608436.000000  608436.000000   
mean   550037.873084    2020.000000       6.500000       3.908107   
std    258935.180575       0.816497       3.452055       5.057889   
min    100018.000000    2019.000000       1.000000       0.000000   
25%    326961.000000    2019.000000       3.750000       0.000000   
50%    550834.000000    2020.000000       6.500000       0.000000   
75%    772194.000000    2021.000000       9.250000       7.200000   
max    999986.000000    2021.000000      12.000000      21.000000   

       NumFlightsWithCompanions     DistanceKM  PointsAccumulated  \
count             608436.000000  608436.000000      608436.000000   
mean                   0.983944    7939.341419         793.777781   
std                    2.003785   10260.421873        1025.918521   
min                    0.000000       0.000000           0.000000   
25%                    0.000000       0.000000           0.000000   
50%                    0.000000     856.400000          85.275000   
75%                    0.900000   15338.175000        1533.712500   
max                   11.000000   42040.000000        4204.000000   

       PointsRedeemed  DollarCostPointsRedeemed  
count   608436.000000             608436.000000  
mean       235.251678                  2.324835  
std        983.233374                  9.725168  
min          0.000000                  0.000000  
25%          0.000000                  0.000000  
50%          0.000000                  0.000000  
75%          0.000000                  0.000000  
max       7496.000000                 74.000000

In [74]:
flightDB.describe(include='object')

YearMonthDate
count         608436
unique            36
top        12/1/2021
freq           16901

o "top" é a moda e "freq" é a frequencia do valor mais frequente

"unique" é a quantidade de valores unicos ((36 datas diferentes pq é o primeiro dia de cada mês durante 3 anos))

"count" é o numero de valores nao nulos

In [75]:
flightDB.replace("", np.nan, inplace=True)
flightDB.isna().sum()

Loyalty#                    0
Year                        0
Month                       0
YearMonthDate               0
NumFlights                  0
NumFlightsWithCompanions    0
DistanceKM                  0
PointsAccumulated           0
PointsRedeemed              0
DollarCostPointsRedeemed    0
dtype: int64

retirar?

In [76]:
flightDB.describe(include='all').T

count unique        top   freq           mean  \
Loyalty#                  608436.0    NaN        NaN    NaN  550037.873084   
Year                      608436.0    NaN        NaN    NaN         2020.0   
Month                     608436.0    NaN        NaN    NaN            6.5   
YearMonthDate               608436     36  12/1/2021  16901            NaN   
NumFlights                608436.0    NaN        NaN    NaN       3.908107   
NumFlightsWithCompanions  608436.0    NaN        NaN    NaN       0.983944   
DistanceKM                608436.0    NaN        NaN    NaN    7939.341419   
PointsAccumulated         608436.0    NaN        NaN    NaN     793.777781   
PointsRedeemed            608436.0    NaN        NaN    NaN     235.251678   
DollarCostPointsRedeemed  608436.0    NaN        NaN    NaN       2.324835   

                                    std       min       25%       50%  \
Loyalty#                  258935.180575  100018.0  326961.0  550834.0   
Year                           0.816497    2019.0    2019.0    2020.0   
Month                          3.452055       1.0      3.75       6.5   
YearMonthDate                       NaN       NaN       NaN       NaN   
NumFlights                     5.057889       0.0       0.0       0.0   
NumFlightsWithCompanions       2.003785       0.0       0.0       0.0   
DistanceKM                 10260.421873       0.0       0.0     856.4   
PointsAccumulated           1025.918521       0.0       0.0    85.275   
PointsRedeemed               983.233374       0.0       0.0       0.0   
DollarCostPointsRedeemed       9.725168       0.0       0.0       0.0   

                                75%       max  
Loyalty#                   772194.0  999986.0  
Year                         2021.0    2021.0  
Month                          9.25      12.0  
YearMonthDate                   NaN       NaN  
NumFlights                      7.2      21.0  
NumFlightsWithCompanions        0.9      11.0  
DistanceKM                15338.175   42040.0  
PointsAccumulated         1533.7125    4204.0  
PointsRedeemed                  0.0    7496.0  
DollarCostPointsRedeemed        0.0      74.0

In [77]:
flightDB.dtypes

Loyalty#                      int64
Year                          int64
Month                         int64
YearMonthDate                object
NumFlights                  float64
NumFlightsWithCompanions    float64
DistanceKM                  float64
PointsAccumulated           float64
PointsRedeemed              float64
DollarCostPointsRedeemed    float64
dtype: object

In [ ]:
# Converter as colunas para integers
# flightDB.NumFlights = flightDB.NumFlights.astype("int")
# flightDB.NumFlightsWithCompanions = flightDB.NumFlightsWithCompanions.astype("int")
# flightDB.PointsRedeemed = flightDB.PointsRedeemed.astype("int")
# flightDB.DollarCostPointsRedeemed = flightDB.DollarCostPointsRedeemed.astype("int")


# ver se é a melhor approach

In [79]:
# flightDB.dtypes

## Relação KM e Pontos

In [84]:
flightDB[["DistanceKM", "PointsAccumulated"]]

DistanceKM  PointsAccumulated
0           9384.0             938.00
1              0.0               0.00
2          14745.0            1474.00
3          26311.0            2631.00
4          19275.0            1927.00
...            ...                ...
608431     30766.5            3076.65
608432         0.0               0.00
608433     18261.0            1826.10
608434         0.0               0.00
608435         0.0               0.00

[608436 rows x 2 columns]

A cada 10km, é 1 ponto mas

nos primeiros 5, arredonda para baixo, sendo todos números inteiros

nos últimos 5, mantém as decimais

In [93]:
flightDB[["PointsRedeemed", "DollarCostPointsRedeemed"]][flightDB.PointsRedeemed > 0]

PointsRedeemed  DollarCostPointsRedeemed
3                 3213                      32.0
29                4638                      46.0
50                4050                      40.0
56                5151                      51.0
58                6244                      62.0
...                ...                       ...
608328            2759                      27.0
608343            4783                      47.7
608374            4127                      40.5
608414            2709                      27.0
608415            4654                      45.9

[35393 rows x 2 columns]

100 points = 1 dollar

arredondando para baixo, sem cêntimos nos dados mais recentes (de cima)

com casas decimais, com cêntimos nos dados mais recentes (de cima)

In [104]:
numeric_variables= flightDB.loc[:,['NumFlights', 'NumFlightsWithCompanions', 'PointsAccumulated', 'PointsRedeemed', 'DollarCostPointsRedeemed']]
numeric_variables.sum()

NumFlights                  2.328622e+06
NumFlightsWithCompanions    5.986672e+05
PointsAccumulated           4.829630e+08
PointsRedeemed              1.431306e+08
DollarCostPointsRedeemed    1.414513e+06
dtype: float64

# Check Duplicate Values

In [122]:
flightDB.duplicated().sum()

np.int64(2903)

In [123]:
flightDB[flightDB.duplicated()]

Loyalty#  Year  Month YearMonthDate  NumFlights  \
3150      263267  2020      6      6/1/2020           0   
3442      584833  2020      6      6/1/2020           0   
5193      351615  2020      5      5/1/2020           0   
5371      979115  2020      5      5/1/2020           0   
5397      951353  2020      5      5/1/2020           0   
...          ...   ...    ...           ...         ...   
607515    950980  2019     12     12/1/2019           0   
607519    951353  2019     12     12/1/2019           0   
607696    960050  2019     12     12/1/2019           0   
608111    981508  2019     12     12/1/2019           0   
608264    990512  2019     12     12/1/2019           0   

        NumFlightsWithCompanions  DistanceKM  PointsAccumulated  \
3150                         0.0         0.0                0.0   
3442                         0.0         0.0                0.0   
5193                         0.0         0.0                0.0   
5371                         0.0         0.0                0.0   
5397                         0.0         0.0                0.0   
...                          ...         ...                ...   
607515                       0.0         0.0                0.0   
607519                       0.0         0.0                0.0   
607696                       0.0         0.0                0.0   
608111                       0.0         0.0                0.0   
608264                       0.0         0.0                0.0   

        PointsRedeemed  DollarCostPointsRedeemed  
3150                 0                       0.0  
3442                 0                       0.0  
5193                 0                       0.0  
5371                 0                       0.0  
5397                 0                       0.0  
...                ...                       ...  
607515               0                       0.0  
607519               0                       0.0  
607696               0                       0.0  
608111               0                       0.0  
608264               0                       0.0  

[2903 rows x 10 columns]

In [124]:
flightDB[flightDB["Loyalty#"] == 263267]


Loyalty#  Year  Month YearMonthDate  NumFlights  \
1092      263267  2020      6      6/1/2020           0   
3150      263267  2020      6      6/1/2020           0   
14057     263267  2020      5      5/1/2020           0   
25441     263267  2020      5      5/1/2020           0   
37425     263267  2020      4      4/1/2020           0   
...          ...   ...    ...           ...         ...   
560775    263267  2019     10     10/1/2019           0   
577675    263267  2019     11     11/1/2019           0   
577676    263267  2019     11     11/1/2019           0   
594576    263267  2019     12     12/1/2019           0   
594577    263267  2019     12     12/1/2019           0   

        NumFlightsWithCompanions  DistanceKM  PointsAccumulated  \
1092                         0.0         0.0                0.0   
3150                         0.0         0.0                0.0   
14057                        0.0         0.0                0.0   
25441                        0.0         0.0                0.0   
37425                        0.0         0.0                0.0   
...                          ...         ...                ...   
560775                       0.0         0.0                0.0   
577675                       0.0         0.0                0.0   
577676                       0.0         0.0                0.0   
594576                       0.0         0.0                0.0   
594577                       0.0         0.0                0.0   

        PointsRedeemed  DollarCostPointsRedeemed  
1092                 0                       0.0  
3150                 0                       0.0  
14057                0                       0.0  
25441                0                       0.0  
37425                0                       0.0  
...                ...                       ...  
560775               0                       0.0  
577675               0                       0.0  
577676               0                       0.0  
594576               0                       0.0  
594577               0                       0.0  

[72 rows x 10 columns]

72 rows iguais, ou seja, todos os 36 unique values estão duplicados.

In [126]:
flightDB.drop_duplicates()

Loyalty#  Year  Month YearMonthDate  NumFlights  \
0         413052  2021     12     12/1/2021           2   
1         464105  2021     12     12/1/2021           0   
2         681785  2021     12     12/1/2021          10   
3         185013  2021     12     12/1/2021          16   
4         216596  2021     12     12/1/2021           9   
...          ...   ...    ...           ...         ...   
608431    999902  2019     12     12/1/2019           7   
608432    999911  2019     12     12/1/2019           0   
608433    999940  2019     12     12/1/2019          14   
608434    999982  2019     12     12/1/2019           0   
608435    999986  2019     12     12/1/2019           0   

        NumFlightsWithCompanions  DistanceKM  PointsAccumulated  \
0                            2.0      9384.0             938.00   
1                            0.0         0.0               0.00   
2                            3.0     14745.0            1474.00   
3                            4.0     26311.0            2631.00   
4                            0.0     19275.0            1927.00   
...                          ...         ...                ...   
608431                       0.0     30766.5            3076.65   
608432                       0.0         0.0               0.00   
608433                       0.9     18261.0            1826.10   
608434                       0.0         0.0               0.00   
608435                       0.0         0.0               0.00   

        PointsRedeemed  DollarCostPointsRedeemed  
0                    0                       0.0  
1                    0                       0.0  
2                    0                       0.0  
3                 3213                      32.0  
4                    0                       0.0  
...                ...                       ...  
608431               0                       0.0  
608432               0                       0.0  
608433               0                       0.0  
608434               0                       0.0  
608435               0                       0.0  

[605533 rows x 10 columns]

In [127]:
flightDB.duplicated().sum()

np.int64(2903)

In [130]:
flightDB2 = flightDB.drop_duplicates()
flightDB2.duplicated().sum()

np.int64(0)

# General Look at the Costumer DataSet

In [4]:
customerDB.shape

(16921, 21)

In [11]:
customerDB.head(10).T

0            1                 2  \
Unnamed: 0                                 0            1                 2   
Loyalty#                              480934       549612            429460   
First Name                           Cecilia        Dayle            Necole   
Last Name                        Householder        Menez            Hannon   
Customer Name            Cecilia Householder  Dayle Menez     Necole Hannon   
Country                               Canada       Canada            Canada   
Province or State                    Ontario      Alberta  British Columbia   
City                                 Toronto     Edmonton         Vancouver   
Latitude                           43.653225    53.544388          49.28273   
Longitude                         -79.383186   -113.49093        -123.12074   
Postal code                          M2Z 4K1      T3G 6Y6           V6E 3D9   
Gender                                female         male              male   
Education                           Bachelor      College           College   
Location Code                          Urban        Rural             Urban   
Income                               70146.0          0.0               0.0   
Marital Status                       Married     Divorced            Single   
LoyaltyStatus                           Star         Star              Star   
EnrollmentDateOpening              2/15/2019     3/9/2019         7/14/2017   
CancellationDate                         NaN          NaN          1/8/2021   
Customer Lifetime Value              3839.14      3839.61           3839.75   
EnrollmentType                      Standard     Standard          Standard   

                                   3               4                5  \
Unnamed: 0                         3               4                5   
Loyalty#                      608370          530508           193662   
First Name                     Queen          Claire         Leatrice   
Last Name                      Hagee         Latting           Hanlin   
Customer Name            Queen Hagee  Claire Latting  Leatrice Hanlin   
Country                       Canada          Canada           Canada   
Province or State            Ontario          Quebec            Yukon   
City                         Toronto            Hull       Whitehorse   
Latitude                   43.653225        45.42873        60.721188   
Longitude                 -79.383186      -75.713364       -135.05684   
Postal code                  P1W 1K4         J8Y 3Z5          Y2K 6R0   
Gender                          male            male             male   
Education                    College        Bachelor         Bachelor   
Location Code               Suburban        Suburban            Rural   
Income                           0.0         97832.0          26262.0   
Marital Status                Single         Married          Married   
LoyaltyStatus                   Star            Star             Star   
EnrollmentDateOpening      2/17/2016      10/25/2017         5/7/2015   
CancellationDate                 NaN             NaN              NaN   
Customer Lifetime Value      3839.75         3842.79          3844.57   
EnrollmentType              Standard  2021 Promotion         Standard   

                                   6           7                 8  \
Unnamed: 0                         6           7                 8   
Loyalty#                      927943      188893            852392   
First Name                       Hue       Nakia            Arlene   
Last Name                    Sellner        Cash         Conterras   
Customer Name            Hue Sellner  Nakia Cash  Arlene Conterras   
Country                       Canada      Canada            Canada   
Province or State            Ontario     Ontario            Quebec   
City                         Toronto     Trenton          Montreal   
Latitude                   43.653225   44.101128          45.50169 

In [10]:
customerDB.tail().T 
#pusemos em trasnposto para ver melhor a informação de todas as features

16916        16917           16918  \
Unnamed: 0                           15           16              17   
Loyalty#                         100012       100013          100014   
First Name                        Ethan        Layla          Amelia   
Last Name                      Thompson        Young         Bennett   
Customer Name            Ethan Thompson  Layla Young  Amelia Bennett   
Country                          Canada       Canada          Canada   
Province or State                Quebec      Alberta   New Brunswick   
City                        Quebec City     Edmonton         Moncton   
Latitude                      46.759733    53.524829       46.051866   
Longitude                    -71.141009  -113.546357      -64.825428   
Postal code                     Y0C 7D6      L3S 9Y3         G2S 2B6   
Gender                             male       female            male   
Education                      Bachelor     Bachelor        Bachelor   
Location Code                  Suburban        Rural           Rural   
Income                              NaN          NaN             NaN   
Marital Status                   Single      Married         Married   
LoyaltyStatus                      Star         Star            Star   
EnrollmentDateOpening         2/27/2019    9/20/2017      11/28/2020   
CancellationDate              2/27/2019    9/20/2017      11/28/2020   
Customer Lifetime Value             NaN          NaN             NaN   
EnrollmentType                 Standard     Standard        Standard   

                                   16919             16920  
Unnamed: 0                            18                19  
Loyalty#                          100015            100016  
First Name                      Benjamin              Emma  
Last Name                         Wilson            Martin  
Customer Name            Benjamin Wilson       Emma Martin  
Country                           Canada            Canada  
Province or State                 Quebec  British Columbia  
City                         Quebec City      Dawson Creek  
Latitude                        46.86297         55.720562  
Longitude                     -71.133444        -120.16009  
Postal code                      B1Z 8T3           M4A 1E4  
Gender                            female            female  
Education                        College            Master  
Location Code                      Urban          Suburban  
Income                               NaN               NaN  
Marital Status                   Married            Single  
LoyaltyStatus                       Star              Star  
EnrollmentDateOpening           4/9/2020         7/21/2020  
CancellationDate                4/9/2020         7/21/2020  
Customer Lifetime Value              NaN               NaN  
EnrollmentType                  Standard          Standard

In [9]:
customerDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16921 entries, 0 to 16920
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               16921 non-null  int64  
 1   Loyalty#                 16921 non-null  int64  
 2   First Name               16921 non-null  object 
 3   Last Name                16921 non-null  object 
 4   Customer Name            16921 non-null  object 
 5   Country                  16921 non-null  object 
 6   Province or State        16921 non-null  object 
 7   City                     16921 non-null  object 
 8   Latitude                 16921 non-null  float64
 9   Longitude                16921 non-null  float64
 10  Postal code              16921 non-null  object 
 11  Gender                   16921 non-null  object 
 12  Education                16921 non-null  object 
 13  Location Code            16921 non-null  object 
 14  Income                

Conclusões da análise anterior: temos valores nulos nas variáveis "Income", "CancellationDate" e "CustomerLifetimeValue"

Podemos dizer faz sentido ter NaN values na "CancellationDate", pois isso significa que há clientes que não abandonaram o programa

Em relação à variável "Income" poderá fazer sentido ter NaN values nos casos em que os clientes não queiram partilhar o seu rendimento pessoal anual. Ou podem tb ser erros de input. (Depende da interpretação)

Para a variável "CustomerLifetimeValue" achamos que não faz sentido existirem NaN values pq mesmo que o cliente não tenha nenhum valor para a empresa o seu CustomerLifetimeValue será 0.

In [13]:
customerDB.dtypes

Unnamed: 0                   int64
Loyalty#                     int64
First Name                  object
Last Name                   object
Customer Name               object
Country                     object
Province or State           object
City                        object
Latitude                   float64
Longitude                  float64
Postal code                 object
Gender                      object
Education                   object
Location Code               object
Income                     float64
Marital Status              object
LoyaltyStatus               object
EnrollmentDateOpening       object
CancellationDate            object
Customer Lifetime Value    float64
EnrollmentType              object
dtype: object

Estes tipos podem estar mal consoante o que considerarmos nas datas (objeto ou inteiro)

In [18]:
customerDB.describe()

Unnamed: 0       Loyalty#      Latitude     Longitude        Income  \
count  16921.000000   16921.000000  16921.000000  16921.000000  16901.000000   
mean    8440.023639  550197.393771     47.174500    -91.814768  37758.038400   
std     4884.775439  259251.503597      3.307971     22.242429  30368.992499   
min        0.000000  100011.000000     42.984924   -135.056840      0.000000   
25%     4210.000000  326823.000000     44.231171   -120.237660      0.000000   
50%     8440.000000  550896.000000     46.087818    -79.383186  34161.000000   
75%    12670.000000  772438.000000     49.282730    -74.596184  62396.000000   
max    16900.000000  999999.000000     60.721188    -52.712578  99981.000000   

       Customer Lifetime Value  
count             16901.000000  
mean               7990.460188  
std                6863.173093  
min                1898.010000  
25%                3979.720000  
50%                5780.180000  
75%                8945.690000  
max               83325.380000

In [20]:
customerDB.describe(include='object').T

count unique                  top   freq
First Name             16921   4941                 Deon     13
Last Name              16921  15404              Salberg      4
Customer Name          16921  16921  Cecilia Householder      1
Country                16921      1               Canada  16921
Province or State      16921     11              Ontario   5468
City                   16921     29              Toronto   3390
Postal code            16921     75              V6E 3D9    917
Gender                 16921      2               female   8497
Education              16921      5             Bachelor  10586
Location Code          16921      3             Suburban   5716
Marital Status         16921      3              Married   9842
LoyaltyStatus          16921      3                 Star   7761
EnrollmentDateOpening  16921   2449             4/3/2015     34
CancellationDate        2310   1260             7/7/2020      8
EnrollmentType         16921      2             Standard  15773

In [21]:
customerDB.describe(include ='all').T

count unique                  top   freq  \
Unnamed: 0               16921.0    NaN                  NaN    NaN   
Loyalty#                 16921.0    NaN                  NaN    NaN   
First Name                 16921   4941                 Deon     13   
Last Name                  16921  15404              Salberg      4   
Customer Name              16921  16921  Cecilia Householder      1   
Country                    16921      1               Canada  16921   
Province or State          16921     11              Ontario   5468   
City                       16921     29              Toronto   3390   
Latitude                 16921.0    NaN                  NaN    NaN   
Longitude                16921.0    NaN                  NaN    NaN   
Postal code                16921     75              V6E 3D9    917   
Gender                     16921      2               female   8497   
Education                  16921      5             Bachelor  10586   
Location Code              16921      3             Suburban   5716   
Income                   16901.0    NaN                  NaN    NaN   
Marital Status             16921      3              Married   9842   
LoyaltyStatus              16921      3                 Star   7761   
EnrollmentDateOpening      16921   2449             4/3/2015     34   
CancellationDate            2310   1260             7/7/2020      8   
Customer Lifetime Value  16901.0    NaN                  NaN    NaN   
EnrollmentType             16921      2             Standard  15773   

                                  mean            std        min        25%  \
Unnamed: 0                 8440.023639    4884.775439        0.0     4210.0   
Loyalty#                 550197.393771  259251.503597   100011.0   326823.0   
First Name                         NaN            NaN        NaN        NaN   
Last Name                          NaN            NaN        NaN        NaN   
Customer Name                      NaN            NaN        NaN        NaN   
Country                            NaN            NaN        NaN        NaN   
Province or State                  NaN            NaN        NaN        NaN   
City                               NaN            NaN        NaN        NaN   
Latitude                       47.1745       3.307971  42.984924  44.231171   
Longitude                   -91.814768      22.242429 -135.05684 -120.23766   
Postal code                        NaN            NaN        NaN        NaN   
Gender                             NaN            NaN        NaN        NaN   
Education                          NaN            NaN        NaN        NaN   
Location Code                      NaN            NaN        NaN        NaN   
Income                      37758.0384   30368.992499        0.0        0.0   
Marital Status                     NaN            NaN        NaN        NaN   
LoyaltyStatus                      NaN            NaN        NaN        NaN   
EnrollmentDateOpening              NaN            NaN        NaN        NaN   
CancellationDate                   NaN            NaN        NaN        NaN   
Customer Lifetime Value    7990.460188    6863.173093    1898.01    3979.72   
EnrollmentType                     NaN            NaN        NaN        NaN   

                               50%        75%        max  
Unnamed: 0                  8440.0    12670.0    16900.0  
Loyalty#                  550896.0   772438.0   999999.0  
First Name                     NaN        NaN        NaN  
Last Name                      NaN        NaN        NaN  
Customer Name                  NaN        NaN        NaN  
Country                        NaN        NaN        NaN  
Province or State              NaN        NaN        NaN  
City                           NaN        NaN        NaN  
Latitude                 46.087818   49.28273  60.721188  
Longitude               -79.383186 -74.596184 -52.712578  
Postal code                    NaN        NaN        NaN  
Gender           

In [22]:
customerDB.replace("", np.nan, inplace=True)
customerDB.isna().sum()

Unnamed: 0                     0
Loyalty#                       0
First Name                     0
Last Name                      0
Customer Name                  0
Country                        0
Province or State              0
City                           0
Latitude                       0
Longitude                      0
Postal code                    0
Gender                         0
Education                      0
Location Code                  0
Income                        20
Marital Status                 0
LoyaltyStatus                  0
EnrollmentDateOpening          0
CancellationDate           14611
Customer Lifetime Value       20
EnrollmentType                 0
dtype: int64

# Comparação de Variáveis

In [28]:
customerDB[["Income", "Customer Lifetime Value"]]

Income  Customer Lifetime Value
0      70146.0                  3839.14
1          0.0                  3839.61
2          0.0                  3839.75
3          0.0                  3839.75
4      97832.0                  3842.79
...        ...                      ...
16916      NaN                      NaN
16917      NaN                      NaN
16918      NaN                      NaN
16919      NaN                      NaN
16920      NaN                      NaN

[16921 rows x 2 columns]

Daqui podemos talvez concluir que 20 NaN values no Income poderão ser os mesmo 20 do Costumer LifeTime Value

In [30]:
customerDB[["EnrollmentDateOpening", "CancellationDate"]]

EnrollmentDateOpening CancellationDate
0                 2/15/2019              NaN
1                  3/9/2019              NaN
2                 7/14/2017         1/8/2021
3                 2/17/2016              NaN
4                10/25/2017              NaN
...                     ...              ...
16916             2/27/2019        2/27/2019
16917             9/20/2017        9/20/2017
16918            11/28/2020       11/28/2020
16919              4/9/2020         4/9/2020
16920             7/21/2020        7/21/2020

[16921 rows x 2 columns]

Mais uma vez os ultimos valores parecem nao fazer sentido, pq o dia de adesão é o mesmo que o dia do cancelamento. Provavelmente teremos de eliminar estas observações pq devem ser erros

!!!! Faz sentido comparar muitas variáveis duas a duas então talvez seja mais fácil realizar os histogramas de cada par de variáveis para comparar mais rapidamente todos os pares e tirar conclusões

# Verificar Duplicados

In [35]:
customerDB_no_name = customerDB.drop(columns=["First Name", "Last Name", "Customer Name"])
customerDB_no_name.duplicated().sum()
#raciocínio: criar uma variável sem os nomes pq já vimos que não há nomes repetidos
#mas a stora avisou que isto podia acontecer mas que as restantes features podiam ser iguais
#e estes nomes diferentes podiam ser um erro de input

np.int64(0)